### DISCUSSION TITLE
Working With API's

### DISCUSSION THREAD

**Discussion by BakrMohamedRBM** [pointfreeco/swift-composable-architecture#1545] [Beta]
i can't find any Solution or how it works with API , Anyone Can Help 

Here my API Service

```
struct NewsService {
    static let shared: NewsService = .init()
    
    private let apiFetcher = APIFetcher()
    
    func getNewsList() -> Effect<BaseResponse<[NewsItemModel]?>, NetworkError> {
        let request = NewsRequests.getNewsList
        return apiFetcher.fetch(request: request, responseClass: BaseResponse<[NewsItemModel]?>.self)
            .eraseToEffect()
    }
    
    
    func getNewsDetails(id: String) -> Effect<BaseResponse<[NewsDetailsModel]?>, NetworkError> {
        let request = NewsRequests.getNewsDetails(id: id)
        return apiFetcher.fetch(request: request, responseClass: BaseResponse<[NewsDetailsModel]?>.self)
            .eraseToEffect()
    }
}
```

and Here is the Feature

```
struct NewsFeature: ReducerProtocol {
    
    struct State: Equatable {
        var isLoading: Bool = true
        var newsList: [NewsItemModel] = []
    }
    
    enum Action: Equatable {
        case onAppear
        case loadedNews(TaskResult<BaseResponse<[NewsItemModel]?>>)
    }
        
    func reduce(into state: inout State, action: Action) -> EffectTask<Action> {
        switch action {
        case .onAppear:
            return .task {
                .loadedNews(
                    TaskResult {
                        NewsService.shared.getNewsList() 
// error Cannot convert value of type 'EffectPublisher<BaseResponse<[NewsItemModel]?>, NetworkError>' to closure result type 'BaseResponse<[NewsItemModel]?>'
                    }
                )
            }
        case .loadedNews(.success(let response)):
            return .none
        case .loadedNews(.failure( let error)):
            return .none
        }
    }
}

```
What am I missing here also tried catchtoeffect or mapping and erasetoeffect is deprecated 

Appreciate your support 

thanks

**Comment by mbrandonw** [comment]
If you are using the Combine interface to `Effect` then you cannot use `TaskResult` or `Effect.task`/`.run`. Instead you return the effect publisher directly, but you must transform it to be in the right shape.

For your example, you need to update your action case to be:

```swift
case loadedNews(Result<BaseResponse<[NewsItemModel]?>, NetworkError>)
```

…and then in the reducer you can do:

```swift
return .onAppear:
  return NewsService.shared.getNewsList() 
    .catchToEffect(Action.loadedNews)
```

Note that the Combine interface to `Effect` is only soft deprecated. It will not currently produce warnings to use it. Someday in the future it will be officially deprecated, and then sometime after that removed. You may want to look into converting to async/await at some point.

Also note that it is not recommended to reach out to uncontrolled dependencies such as `NewsService.shared`. Instead you should leverage the dependency management system to access your API client in a controllable manner. We have an [article](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/dependencymanagement) about this topic that you may find interesting.

**Reply by BakrMohamedRBM** [reply]
Many thanks for the advice and help I will take your advice and read the article 
you do amazing here thanks <3
